src/Pos/Delegation/Listeners.hs

```diff
@@ -0,0 +1,157 @@
+import           Universum
+
+
+import           Pos.Binary.Communication   ()
+import           Pos.Communication.Protocol (ListenerSpec, OutSpecs, SendActions (..),
+                                             listenerOneMsg, mergeLs, oneMsgH, toOutSpecs)
+import           Pos.Context                (getNodeContext, ncBlkSemaphore, ncNodeParams,
+                                             npPropagation)
+import           Pos.Delegation.Logic       (ConfirmPskLightVerdict (..),
+                                             PskHeavyVerdict (..), PskLightVerdict (..),
+                                             processConfirmProxySk, processProxySKHeavy,
+                                             processProxySKLight)
+import           Pos.Delegation.Methods     (sendProxyConfirmSK, sendProxyConfirmSKOuts,
+                                             sendProxySKHeavy, sendProxySKHeavyOuts,
+                                             sendProxySKLight, sendProxySKLightOuts)
+import           Pos.Delegation.Types       (ConfirmProxySK (..), SendProxySK (..))
+import           Pos.DHT.Model              (sendToNeighbors)
+import           Pos.Types                  (ProxySKLight)
+import           Pos.WorkMode               (WorkMode)
+
+-- | Listeners for requests related to delegation processing.
+delegationListeners
+    :: WorkMode ssc m
+    => ([ListenerSpec m], OutSpecs)
+delegationListeners = mergeLs
+    [ handleSendProxySK
+    , handleConfirmProxySK
+    --, handleCheckProxySKConfirmed
+    ]
+
+delegationStubListeners
+    :: WithLogger m
+    => ([ListenerSpec m], OutSpecs)
+delegationStubListeners = mergeLs
+    [ stubListenerOneMsg (Proxy :: Proxy SendProxySK)
+    , stubListenerOneMsg (Proxy :: Proxy ConfirmProxySK)
+    --, stubListenerOneMsg (Proxy :: Proxy CheckProxySKConfirmed)
+    ]
+
+----------------------------------------------------------------------------
+-- PSKs propagation
+----------------------------------------------------------------------------
+
+-- | Handler 'SendProxySK' event.
+handleSendProxySK
+    :: forall ssc m.
+       (WorkMode ssc m)
+    => (ListenerSpec m, OutSpecs)
+handleSendProxySK = listenerOneMsg outSpecs $
+    \_ _ sendActions (pr :: SendProxySK) -> handleDo sendActions pr
+  where
+    handleDo sendActions req@(SendProxySKHeavy pSk) = do
+        logDebug $ sformat ("Got request to handle heavyweight psk: "%build) pSk
+        verdict <- processProxySKHeavy pSk
+        logDebug $ sformat ("The verdict for cert "%build%" is: "%shown) pSk verdict
+        doPropagate <- npPropagation . ncNodeParams <$> getNodeContext
+        if | verdict == PHIncoherent -> do
+               -- We're probably updating state over epoch, so leaders
+               -- can be calculated incorrectly.
+               blkSemaphore <- ncBlkSemaphore <$> getNodeContext
+               void $ liftIO $ readMVar blkSemaphore
+               handleDo sendActions req
```

Comment:
⚠️ This `readMVar` and then `handleDo` is concerning.

If `handleDo` is atomic -- i.e. it does not handle resources that other threads can see, then this code is OK. However, if it's not atomic, then this `readMVar` is multiple wake up and all the threads waiting will wake up. If that's not desirable, it should be a `takeMVar`. 

Need to understand the wider scope of this function to confirm that.